### DISCUSSION TITLE
Multiple CMDT with WHERE clauses

### DISCUSSION THREAD

**Discussion by pyao-bwc** [jamessimone/apex-rollup#417] [Q&A] [answered]
Hi @jamessimone , we'd like some of our rollups to run on schedule (some date-based rollups will change without corresponding dml).

To start, I'm trying to schedule these two:
![image](https://user-images.githubusercontent.com/23371412/223516822-94ee4171-81f0-451e-936d-0d5e268752f4.png)

They have different child objects, and one has a where clause and the other doesn't.

When run individually, they complete without exception. But when run together (via [this snippet](https://gist.github.com/pyao-bwc/94cc1d50f86a30369e6a56a08edcdf9b), if it's relevant), we get this exception:
```
System.SObjectException: Invalid field Next_Schedule_Z_Status__c for Schedule_Z_Subscription__c

Class.RollupEvaluator.WhereFieldCondition.getFieldValue: line 696, column 1
Class.RollupEvaluator.WhereFieldCondition.equals: line 608, column 1
Class.RollupEvaluator.AndConditionalGrouping.conditionallyEquals: line 552, column 1
Class.RollupEvaluator.ConditionalGrouping.equals: line 510, column 1
Class.RollupEvaluator.WhereFieldEvaluator.matches: line 250, column 1
Class.RollupCalculator.winnowItems: line 306, column 1
Class.RollupCalculator.performRollup: line 173, column 1
Class.RollupAsyncProcessor.getCalculator: line 1255, column 1
Class.RollupAsyncProcessor.getUpdatedLookupItemsByRollup: line 1160, column 1
Class.RollupAsyncProcessor.process: line 659, column 1
Class.RollupAsyncProcessor.performWork: line 332, column 1
Class.RollupAsyncProcessor.QueueableProcessor.execute: line 421, column 1
```
Here's the corresponding log: 
[apex-07L76000005Zr0aEAC.log](https://github.com/jamessimone/apex-rollup/files/10912901/apex-07L76000005Zr0aEAC.log)

Next_Schedule_Z_Status__c is in the WHERE clause for the first CMDT, and Schedule_Z_Subscription__c is the child object of the second CMDT.

I read through [this post](https://github.com/jamessimone/apex-rollup/discussions/321), but that seemed different (and fixed). I started stepping through it and got as far as ``RollupAsyncProcessor.populateObjectFields``, when I thought I'd ask for your help!

Are there other things I could do to debug this?

**Comment by jamessimone** [comment, marked as answer]
@pyao-bwc Thanks for sharing that snippet, super helpful to see what you're doing there. I should be able to reproduce this issue based on what you've provided. I haven't had somebody asking about scheduling a multi-child item rollup yet! 

To be honest, I'm surprised this isn't already working but I should be able to fix the underlying issue based on the info you've provided. I would assume this fix will be in the  _next_ version (`v1.5.53`) as I'd like to get the fixes I'm making for #414 fixed for you definitively.

**Reply by pyao-bwc** [reply]
Super, thanks again. And if I'm straying off the beaten path, let me know and I'll follow the more common pattern!

**Reply by jamessimone** [reply]
There's always another use case to support - what I love is to see this kind of thinking, because it indicates to me an area of improvement within the package!

**Reply by jamessimone** [reply]
@pyao-bwc interesting - I went to create a failing test for this but I _think_ I've discovered how you can get this working (based on your snippet):

```diff
            List<RollupFlowFullRecalcDispatcher.FlowInput> inputs = new List<RollupFlowFullRecalcDispatcher.FlowInput>();
+         RollupFlowFullRecalcDispatcher.FlowInput input = new RollupFlowFullRecalcDispatcher.FlowInput();
+         input.rollupDeveloperNames = '';
            for (Rollup__mdt rollup : localRollupMetadata) {
-                RollupFlowFullRecalcDispatcher.FlowInput input = new RollupFlowFullRecalcDispatcher.FlowInput();
+                input.rollupDeveloperNames += rollup.DeveloperName + ',';
-                inputs.add(input);
            }
+          input.rollupDeveloperNames = input.rollupDeveloperNames.removeEnd(',');
+          inputs.add(input);  
```

Let me know if that makes sense and if that indeed fixes the issue. Because this method is traditionally called from Flow, it's really set up to be a comma-separated list of metadata developer names within a single instance of `RollupFlowFullRecalcDispatcher.FlowInput`.

**Reply by pyao-bwc** [reply]
Oh no - I'm sorry about that, I'll give that a try later today and let you know!

**Reply by jamessimone** [reply]
No worries whatsoever - and my suggestion may not make a difference at all, but I want to see if it does since it would seem to exactly mirror the passing test I have for your situation. This class, `RollupFlowFullRecalcDispatcher`, was a community addition and as such I haven't spent as much time on it as I have with most of the rest of the codebase, so it's good for me to also learn about any peculiarities specific to this class.

**Reply by pyao-bwc** [reply]
That makes the stacktrace go away! There's still something going on, maybe because I'm running it via ``RollupFlowFullRecalcDispatcher`` and also via trigger, or maybe one of the rollups conflicts with another, but I'll dig into it more. Thanks again!

**Reply by jamessimone** [reply]
Ok great. Let me know where that leaves you; I'll continue to work with you to ensure everything's functioning as planned.

**Reply by pyao-bwc** [reply]
Hmm - I stand corrected, I was running nine rollups and not looking at the right log.

When I run ``RollupFlowFullRecalcDispatcher`` with two rollups with different child objects (I updated [the gist](https://gist.github.com/pyao-bwc/94cc1d50f86a30369e6a56a08edcdf9b) to make that easier), I can get that stacktrace again:
[apex-07L76000005a2GYEAY.log](https://github.com/jamessimone/apex-rollup/files/10927745/apex-07L76000005a2GYEAY.log)

But I can run other pairs of rollups with different child objects successfully, for instance, like this:
[apex-07L76000005a2N4EAI.log](https://github.com/jamessimone/apex-rollup/files/10927782/apex-07L76000005a2N4EAI.log)

**Reply by jamessimone** [reply]
I'll take a look in the morning and see what's going on. I had pushed up a commit for a preliminary stop-gap fix that will probably fix the issue downstream, but I'd prefer to be able to actually reproduce the issue at its source, which is why I was confused that the test that I wrote for two different children objects being rolled up from that code path worked no problem.

It's my hope that by cross-analyzing your working case against the non-working one will reveal some additional flaw with my new test to help further shed light on the actual issue.

**Reply by jamessimone** [reply]
I was able to start down this thread this morning - I should have more time to continue to investigate this afternoon. Based on your working versus not-working example, it's interesting that the working example only has one grandparent rollup. More to come.

**Reply by jamessimone** [reply]
Ok I now see what the actual problem is, those logs were extremely helpful in getting to the bottom of this! I expect I'll be able to get this patched up and released by tomorrow.

**Reply by jamessimone** [reply]
@pyao-bwc this should be fixed with the release of #419. Let me know!

**Reply by pyao-bwc** [reply]
Thanks @jamessimone , I'll get back to you today!

**Reply by pyao-bwc** [reply]
Ha ha! After [fixing](https://gist.github.com/pyao-bwc/94cc1d50f86a30369e6a56a08edcdf9b/revisions) my gist, this is looking great! Thanks for that fix!

**Reply by jamessimone** [reply]
Perfect - glad to hear it

**Reply by pyao-bwc** [reply]
Hi @jamessimone , it looks like issue #411 which #412 fixed and we verified is back. One of our apex tests failed, and I can reproduce the issue manually as well, with this debug log:
[apex-07L0400000A5sw5EAB.log](https://github.com/jamessimone/apex-rollup/files/10957941/apex-07L0400000A5sw5EAB.log)

Looks like v1.5.49 doesn't have this issue for us. Can you let me know if I can do anything else to help debug it? Thanks!

**Reply by pyao-bwc** [reply]
Oh wait! I'm sorry @jamessimone , the test passes in another sandbox with 1.5.53, so I must have forgotten to commit something. I'll see what's different between sandboxes, it's probably a false alarm!

**Reply by jamessimone** [reply]
Ok whew. Keep me posted!

**Reply by pyao-bwc** [reply]
Hi @jamessimone , well, unfortunately, I couldn't figure out... it's surprising!

I have two sandboxes: ``peter`` where I've been working out of which has the passing test and ``peter2`` which is freshly copied from prod (plus new custom metadata types) where the test fails.

Here's the log for the test that passes in ``peter``:
[all-rollups-enabled-and-tests-pass.txt](https://github.com/jamessimone/apex-rollup/files/10963639/all-rollups-enabled-and-tests-pass.txt)

Here's the log for the same test that fails in ``peter2``:
[all-rollups-enabled-and-test-fails.txt](https://github.com/jamessimone/apex-rollup/files/10963638/all-rollups-enabled-and-test-fails.txt)

I can disable one of the rollups in ``peter2`` and it still fails:
[one-rollup-disabled-and-test-fails.txt](https://github.com/jamessimone/apex-rollup/files/10963641/one-rollup-disabled-and-test-fails.txt)

But if I disable another one of the rollups in ``peter2``, it passes!
[two-rollups-disabled-but-test-passes.txt](https://github.com/jamessimone/apex-rollup/files/10963643/two-rollups-disabled-but-test-passes.txt)

So - it seems like the order that apex-rollup processes rollups might matter? I'm not sure why the RollupCalcItemReplacer doesn't get to do it's thing when the tests fail... I'll try to step through it tomorrow, but you might have some ideas as well!

**Reply by jamessimone** [reply]
I'll take a look through the logs and let you know! Just a head's up that it might be a day or two, but I really appreciate you providing those logs and I'll get to this as soon as I'm able

**Reply by pyao-bwc** [reply]
Sounds good, thanks again!
